### ISSUE TITLE
pmd v7 not supported

### ISSUE THREAD

**Issue by xixiaofinland** [nvimtools/none-ls.nvim#47]
### Issues

- [X] I have checked [existing issues](https://github.com/jose-elias-alvarez/null-ls.nvim/issues) and there are no existing ones with the same request.

### Feature description

pmd version7 code analyzer now switches to use "pmd check" as the main command instead of "pmd", thus null-ls throws a "not recognized parameter" error.
Adding "check" into the `extra_args` table doesn't work.

### Help

Yes, but I don't know how to start. I would need guidance

### Implementation help

I locate the cmd is constructed here: https://github.com/nvimtools/none-ls.nvim/blob/ef09f14eab78ca6ce3bee1ddc73db5511f5cd953/lua/null-ls/helpers/make_builtin.lua#L101

but not sure how to change from "pmd" to "pmd check"

**Comment by jakenvac** [comment]
Hi, thanks for raising an issues. As this is a community maintained project it would be amazing if you could have a go at fixing it yourself. I'd be more than happy to provide whatever guidance I can.

The first place you should take a look is here:
https://github.com/nvimtools/none-ls.nvim/blob/ef09f14eab78ca6ce3bee1ddc73db5511f5cd953/lua/null-ls/builtins/diagnostics/pmd.lua#L104

This is the specific builtin for the PMD diagnostics feature. The `cmd` field is the root command and then `args` is where any subcommands/flags etc. should go. So I'd imagine adding `check` to the args should at least get you moving in the right direction.

Typically, any builtin is configured in its respective type (Formatter, Diagnostic etc.) directory here:
https://github.com/nvimtools/none-ls.nvim/tree/main/lua/null-ls/builtins
 

If you'd like any extra help, please could provide a link to the pmd docs for this command and a java code snippet that you'd expect PMD to provide some diagnostics for. This way those of us that are unfamiliar with java and its tools (me 😆) should still be able to assist you.

Feel free to tag me directly if you're struggling, I'm more likely to see the notification that way.

**Comment by xixiaofinland** [comment]
@jakenvac Thanks Jake for your kind help!
In pmd v6, which null-ls supports, the command is: `**pmd**` + parameters, like below
```
pmd --format json --rulesets apex_ruleset.xml  --dir ./nebula-logger
```

in pmd v7, the command is `**pmd check**` + parameters
```
pmd --format json --rulesets apex_ruleset.xml  --dir ./nebula-logger
```

In [my null-ls configuration](https://github.com/xixiaofinland/dotfiles/blob/8734c9167ddd750392697e6f9a0db91abc6de72e/.config/nvim/lua/plugins/null_ls.lua#L13),  I attempt to add `check` as part of the `extra_args` to form a `pmd check`, but it doesn't work probably because `check` is not a real parameter but part of the main command.

So the question boils down to, how to let none-ls to use `pmd check` instead of `pmd` .

**Comment by xixiaofinland** [comment]
Here is the error in nvim none-ls, as we can see it doesn't invoke `pmd check`
![Image 002](https://github.com/nvimtools/none-ls.nvim/assets/13655323/7b1bb5e6-b87c-419c-9a44-60dbf8fcc3c4)

it is the same error in console, when I run `pmd` instead of `pmd check`
![Image 003](https://github.com/nvimtools/none-ls.nvim/assets/13655323/51bcc5d6-20f8-430c-b24e-bb5d007d4063)

**Comment by jakenvac** [comment]
Hm, it's strange to me that adding `check` into the args doesn't work. While it's a subcommand of pmd it's still technically an argument. We've quite a few integrations that call subcommands with the args.

What error, if any, do you get when adding check to the args?

**Comment by xixiaofinland** [comment]
> Hm, it's strange to me that adding `check` into the args doesn't work. While it's a subcommand of pmd it's still technically an argument. We've quite a few integrations that call subcommands with the args.
> 
> What error, if any, do you get when adding check to the args?

nvim shows exactly the same error as my first screenshot above, as if the `check` is ignored totally.

I suspect the `extra_args` demands the parameter to start with `-` or `--`? In addition, parameters have no order while `check` has to strictly follow after `pmd`, not sure if this matters.
@jakenvac

**Comment by xixiaofinland** [comment]
![Image 005](https://github.com/nvimtools/none-ls.nvim/assets/13655323/03f4b65a-0978-415a-a1e2-ce36946418ea)
When I print `require'null-ls'.builtins.diagnostics.pmd` I got this. @jakenvac

**Comment by xixiaofinland** [comment]
If I try manually overwrite `command` to `pmd check` in my setting, Despite no error, null-ls isn't happy either

![Image 006](https://github.com/nvimtools/none-ls.nvim/assets/13655323/5e41b49c-fc10-46e0-b378-aae1f2370e01)

![Image 007](https://github.com/nvimtools/none-ls.nvim/assets/13655323/f1a57298-560c-4ac1-98de-22e5632f298c)

**Comment by xixiaofinland** [comment]
@jakenvac  could you give examples of other tools in none-ls using subcommands successfully?

**Comment by xixiaofinland** [comment]
As `check` in `extra_args` not working, I created the `args` and put `check` into it. I'm not sure if this is a good practice, but it works. Here is my change [in dotfiles](https://github.com/xixiaofinland/dotfiles/blob/c2f82c8a8f2e5ff69938e35fe62d435607b75369/.config/nvim/lua/plugins/null_ls.lua#L15).

Now at the beginning of the file it still shows an odd error, but analyzed results show correctly (like in line 8)
![Image 008](https://github.com/nvimtools/none-ls.nvim/assets/13655323/3a81ac11-7ad6-4217-a33e-fe76d06b7f7e)

**Comment by jakenvac** [comment]
Could you provide a code sample to run pmd against and a link to the pmd docs? I'll see if I can recreate your issue and debug it.

**Comment by xixiaofinland** [comment]
> Could you provide a code sample to run pmd against and a link to the pmd docs? I'll see if I can recreate your issue and debug it.

Not that I don't want to, it's a salesforce related stuff that has highly custom dotfiles. kinda difficult to reproduce in a clean setup.
Since it is working now (only the odd error at the beginning, it shows the rules already.
IMO, we can treat it as case resolved unless Java/Salesforce developers complains with it in the future :).
Thanks Jake and happy Xmas!

**Comment by FedeAbella** [comment]
@xixiaofinland I don't know if you've completely solved this since closing, but with the stable release of pmd 7, I managed to make it work with no error messages. The null-ls doc says we need a wrapper `pmd` script in `$PATH`, since pmd v6.* used to provide only a `run.sh` script, and gives an example of a wrapper script that can be used:
```
#!/usr/bin/env bash
path/to/pmd/bin/run.sh pmd "$@"
```

With pmd 7, this worked just fine for me just by changing the wrapper to:
```
#!/usr/bin/env bash
path/to/pmd/bin/pmd check "$@"
```

**Comment by xixiaofinland** [comment]
@FedeAbella Thanks, your solution indeed works! The only downside is the `pmd` is now `pmd check`, which is not a big deal :)

**Comment by ognjen-vuceljic** [comment]
@xixiaofinland  @FedeAbella 

Thanks for the insightful conversation.

I am not sure if this thread is still active, or if we can post messages, but I'll give it a go :) 

I've followed this thread closely, and applied all the suggested fixes for `pmd v7`. Seems like `pmd check` is indeed running against the directory with SF classes, however I get the following error:

![image](https://github.com/user-attachments/assets/c3bc69f7-318f-45ef-9955-a3be1f746e3d)


Based on my limited knowledge, I would guess that PMD check generated a fairly large JSON that is not readable (aka out of memory) anymore. I tried excluding some classes from PMD checks, and this number from screenshot gets down, but the error still remains. I do have to note that repo I'm trying to use PMD against is huge, with a ton of classes.

Appreciating in advance any advice and suggestions.

Have a great day!
